lapack/gonum/dlatbs.go

```diff
@@ -0,0 +1,449 @@
+		bi.Dtbsv(uplo, trans, diag, n, kd, ab, ldab, x, 1)
+		// Scale the column norms by 1/tscal for return.
+		if tscal != 1 {
+			bi.Dscal(n, 1/tscal, cnorm, 1)
+		}
+		return 1
+	}
+
+	// Use a Level 1 BLAS solve, scaling intermediate results.
+
+	scale = 1
+	if xMax > bignum {
+		// Scale x so that its components are less than or equal to bignum in
+		// absolute value.
+		scale = bignum / xMax
+		bi.Dscal(n, scale, x, 1)
+		xMax = bignum
+	}
+
+	if noTran {
+		// Solve A * x = b.
+		for j := jFirst; j != jLast; j += jInc {
+			// Compute x[j] = b[j] / A[j,j], scaling x if necessary.
+			xj := math.Abs(x[j])
+			tjjs := tscal
+			if diag == blas.NonUnit {
+				tjjs *= ab[j*ldab+maind]
+			}
+			tjj := math.Abs(tjjs)
+			switch {
+			case tjj > smlnum:
+				// smlnum < abs(A[j,j])
+				if tjj < 1 && xj > tjj*bignum {
+					// Scale x by 1/b[j].
+					rec := 1 / xj
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+					xMax *= rec
+				}
+				x[j] /= tjjs
+				xj = math.Abs(x[j])
+			case tjj > 0:
+				// 0 < abs(A[j,j]) <= smlnum
+				if xj > tjj*bignum {
+					// Scale x by (1/abs(x[j]))*abs(A[j,j])*bignum to avoid
+					// overflow when dividing by A[j,j].
+					rec := tjj * bignum / xj
+					if cnorm[j] > 1 {
+						// Scale by 1/cnorm[j] to avoid overflow when
+						// multiplying x[j] times column j.
+						rec /= cnorm[j]
+					}
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+					xMax *= rec
+				}
+				x[j] /= tjjs
+				xj = math.Abs(x[j])
+			default:
+				// A[j,j] == 0: Set x[0:n] = 0, x[j] = 1, and scale = 0, and
+				// compute a solution to A*x = 0.
+				for i := range x[:n] {
+					x[i] = 0
+				}
+				x[j] = 1
+				xj = 1
+				scale = 0
+				xMax = 0
+			}
+
+			// Scale x if necessary to avoid overflow when adding a multiple of
+			// column j of A.
+			if xj > 1 {
+				rec := 1 / xj
+				if cnorm[j] > (bignum-xMax)*rec {
+					// Scale x by 1/(2*abs(x[j])).
+					rec *= 0.5
+					bi.Dscal(n, rec, x, 1)
+					scale *= rec
+				}
+			} else if xj*cnorm[j] > bignum-xMax {
+				// Scale x by 1/2.
+				bi.Dscal(n, 0.5, x, 1)
+				scale *= 0.5
+			}
+
+			if uplo == blas.Upper {
+				if j > 0 {
+					// Compute the update
+					//  x[max(0,j-kd):j] := x[max(0,j-kd):j] - x[j] * A[max(0,j-kd):j,j]
+					jlen := min(j, kd)
+					if jlen > 0 {
+						bi.Daxpy(jlen, -x[j]*tscal, ab[(j-jlen)*ldab+jlen:], kld, x[j-jlen:], 1)
+					}
+					i := bi.Idamax(j, x, 1)
+					xMax = math.Abs(x[i])
+				}
+			} else if j < n-1 {
+				// Compute the update
+				//  x[j+1:min(j+kd,n)] := x[j+1:min(j+kd,n)] - x[j] * A[j+1:min(j+kd,n),j]
+				jlen := min(kd, n-j-1)
+				if jlen > 0 {
+					bi.Daxpy(jlen, -x[j]*tscal, ab[(j+1)*ldab+kd-1:], kld, x[j+1:], 1)
+				}
+				i := j + 1 + bi.Idamax(n-j-1, x[j+1:], 1)
+				xMax = math.Abs(x[i])
+			}
+		}
+	} else {
+		// Solve A^T * x = b.
+		for j := jFirst; j != jLast; j += jInc {
+			// Compute x[j] = b[j] - sum A[k,j]*x[k].
+			//                       k!=j
+			xj := math.Abs(x[j])
+			tjjs := tscal
```

Comment:
This few lines is conditional in the Fortran. I think I see, it's calculated conditionally there but in two places. This is a hoist. Is that correct?

Reply:
Correct, this is a hoist and hopefully is also clearer, otherwise the logic is quite ugly:
```
DO 160 J = JFIRST, JLAST, JINC
    USCAL = TSCAL
    IF( CNORM( J ).GT.( BIGNUM-XJ )*REC ) THEN
*       Initialize and use TJJS
*       Sometimes modify USCAL
    END IF
* ...
    IF( USCAL.EQ.TSCAL ) THEN
*       Initialize and use TJJS
    ELSE
*       Use TJJS - understanding that TJJS has indeed been initialized here
*       doesn't come immediately.
    END
160 CONTINUE
```
I see no benefit of the logic in the reference.

Reply:
Thanks